mlflow/models/signature.py

```diff
@@ -0,0 +1,449 @@
+                raise TensorsNotSupportedException("Column '{0}' has shape {1}".format(x, y.shape))
+
+        if all([_is_scalar(x) for x in input_example.values()]):
+            input_example = pd.DataFrame([input_example])
+        else:
+            input_example = pd.DataFrame.from_dict(input_example)
+    elif isinstance(input_example, list):
+        for i, x in enumerate(input_example):
+            if isinstance(x, np.ndarray) and len(x.shape) > 1:
+                raise TensorsNotSupportedException("Row '{0}' has shape {1}".format(i, x.shape))
+        if all([_is_scalar(x) for x in input_example]):
+            input_example = pd.DataFrame([input_example])
+        else:
+            input_example = pd.DataFrame(input_example)
+    elif isinstance(input_example, np.ndarray):
+        if len(input_example.shape) > 2:
+            raise TensorsNotSupportedException("Input array has shape {}".format(
+                input_example.shape))
+        input_example = pd.DataFrame(input_example)
+    elif not isinstance(input_example, pd.DataFrame):
+        raise TypeError("Unexpected type of input_example. Expected one of "
+                        "(pandas.DataFrame, numpy.ndarray, dict, list), got {}".format(
+                          type(input_example)))
+
+    example_filename = "input_dataframe_example.json"
+    res = input_example.to_dict(orient="split")
+    # Do not include row index
+    del res["index"]
+    if all(input_example.columns == range(len(input_example.columns))):
+        # No need to write default column index out
+        del res["columns"]
+
+    with open(os.path.join(path, example_filename), "w") as f:
+        json.dump(res, f, cls=NumpyEncoder)
+    return example_filename
+
+
+def _base64decode(x):
+    return base64.decodebytes(x.encode("ascii"))
+
+
+def dataframe_from_json(path_or_str, schema: Schema = None,
+                        pandas_orient: str = "split") -> pd.DataFrame:
+    """
+    Parse json into pandas.DataFrame. User can pass schema to ensure correct type parsing and to
+    make any necessary conversions (e.g. string -> binary for binary columns).
+
+    :param path_or_str: Path to a json file or a json string.
+    :param schema: Mlflow schema used when parsing the data.
+    :param pandas_orient: pandas data frame convention used to store the data.
+    :return: pandas.DataFrame.
+    """
+    if schema is not None:
+        dtypes = dict(zip(schema.column_names(), schema.column_types()))
+        df = pd.read_json(path_or_str, orient=pandas_orient, dtype=dtypes)[schema.column_names()]
+        binary_cols = [i for i, x in enumerate(schema.column_types()) if x == DataType.binary]
+
+        for i in binary_cols:
+            col = df.columns[i]
+            df[col] = np.array(df[col].map(_base64decode), dtype=np.bytes_)
+            return df
+    else:
+        return pd.read_json(path_or_str, orient=pandas_orient, dtype=False)
+
+
+def _map_numpy_dtype(col: np.ndarray) -> DataType:
```

Comment:
Why is this called map as opposed to `_infer_numpy_dtype`? Maybe just a doc comment would be helpful.

Reply:
good question, there is no reason for that! I'll rename it.